### ISSUE TITLE
Major breaking changes while keeping the same package

### ISSUE THREAD

**Issue by masc3d** [EsotericSoftware/kryo#650]
reading migration guide.

> the best approach is to load the old data somehow, then write it with the newer version. For example, by juggling classloaders the data could be loaded with v2, then written with v5.

I fail to see how this is the best or even a good approach. Could you please elaborate?

As far as I see it's best practice for java libs in general to simply move to a new package for major / breaking updates like this eg. `com.esotericsoftware.kryo.v5`, so everybody is happy with smooth migrations.

**Comment by NathanSweet** [comment]
I don't believe it is common practice with Java libraries to rename packages for every binary incompatible version of a library. I do agree a version in the package could make it easier to migrate data. Dropping in a different major version Kryo JAR is never going to "just work" with old data, though that isn't the only use case. It may be that you don't have persistent data (eg, only short lived network communication) or you are starting over without existing data.

Does anyone have thoughts on renaming Kryo v5 classes to a `v5` package? @magro @romix @serverperformance ?

**Comment by masc3d** [comment]
> I don't believe it is common practice with Java libraries to rename packages for every binary incompatible version of a library

not every, but when it makes sense.
and it does on major breaking api changes and even more so on breaking formats.

many of modern java libs do it eg [rxjava](https://github.com/ReactiveX/RxJava), [retrofit](https://github.com/square/retrofit), [swagger](https://github.com/swagger-api/swagger-core) just to name a few.

in a distributed environment where kryo is protocol, supporting v4 and v5 side-by-side is at least temporarily required for a smooth transition.

**Comment by magro** [comment]
As I see it a common motivation for different package names is to prevent a diamond dependency problem. Assuming an application uses different higher level libraries that are using / working with different versions of an underlying lib, updating one of the higher level libs can cause compatibility issues with the underlying lib. Let's take netty as an example: if an application uses a cassandra driver (which uses netty) and async-http-client (using netty as well) it helps if netty changes its package name _and_ group/artifact id on major updates, because the application then can update the cassandra driver or async-http-client independently without running into conflicts with netty (because the high level libs in fact do no longer depend on two different libraries). Another solution for this is shading, i.e. the cassandra driver or async-http-client could come with a shaded netty. In case of kryo we had this issue with asm, why we provided a kryo artifact with a shaded asm. 
In my view that diamond dependency problem is the most common motivation for changing the package name and artifact id on breaking changes. 

For higher level libs only and directly used by application code in general I don't see much value for changing the package and artifact id. 

For kryo changing the package and artifact id would allow applications to use different versions of kryo at the same time, e.g. to deserialize old persisted data with the old kryo and serialize new data with the new kryo. Because kryo is not recommended for persisting serialized data, I'd hope that there are not many users with this need.

All users however would be forced to update their application code interacting with kryo, because of the changed package name (now there are more binary incompatibilities - because kryo was changed to be 100% binary incompatible). 

Thinking about kryo in a distributed system where kryo is used for serialization of messages, I'd expect that application code is updated to the new kryo, and during a rolling deployment the only requirement is serialization (forward and backward) compatibility - so that old and new nodes running different versions of kryo can both deserialize data that was serialized with the older/newer kryo. In that case I don't see no value in changing the package and artifact id - because this only affects binary compatibility but not serialization compatibility. I.e. in the case of broken serialization compatibility, more machinery would be needed to support rolling updates. 

In summary, in the context of applications using kryo directly, I see a tradeoff between supporting users using kryo for persisting data and forcing all users to change their code because kryo was made 100% binary incompatible. 
And I don't know how often kryo is used as a transitive dependency, so that the diamond dependency problem could occur.

**Comment by magro** [comment]
Maybe we could/should provide an _additional_ artifact "kryo5" which comes with renamed package name. Then users could choose their tradeoff. 

Technically this might be possible by adding a module depending on kryo and shading it to the v5 package.

**Comment by johnynek** [comment]
I would really love it if we could not break binary compatibility *ever* and use the approach of a version package name and new artifact name as discussed in this thread.

I maintain: https://github.com/twitter/chill which is a transitive dependency of spark, storm, scalding, summingbird and probably others I don't know about. If we introduce a breaking change in kryo, this becomes really hard to upgrade since some projects depend on more than one of those things above, so they have to be upgraded one at a time on the tree and once all upstreams are upgraded, then you can move forward.

Since projects like Storm and Spark upgrade very slowly, incompatible changes make this very challenging, especially in projects with large transitive classpaths (e.g. many industrial use cases).

That said, long term I would personally really rather move away from dynamic serialization and more to static, codegen or macro based serializers, however, until that day (if it ever really arrives), kryo is still quite useful, and a project I'm quite grateful for.

**Comment by masc3d** [comment]
> Maybe we could/should provide an _additional_ artifact "kryo5" which comes with renamed package name. Then users could choose their tradeoff.

As updating a package name on consumer migration is really just a minor I don't see much benefit in maintaining two packages though. 

Instead it will make it more difficult for supplemental libraries eg [kryo-serializers](https://github.com/magro/kryo-serializers) which would need to follow suit.

**Comment by magro** [comment]
> Instead it will make it more difficult for supplemental libraries eg kryo-serializers which would need to follow suit.

For kryo-serializers it wouldn't make a difference if there'd be the default unversioned artifact as well (assuming that kryo-serializers uses the versioned artifact).

**Comment by johnynek** [comment]
I'm not seeing the benefit of having both the v5 versioned package/artifact and without.

As soon as someone decides to use the v5 without the version and another person stays on previous, you are hosed and the existence of the v5 didn't help you.

A depends on B and both depend on kryo. B upgrades but chooses to stick with the old name. A does not yet upgrade but is on the classpath with B. Now you have breakage right?

What am I missing?

**Comment by magro** [comment]
Not sure if I'm understanding you correctly. What I missed to make explicit (in the PR) is that the 
the produced/published versioned artifact/pom (kryo5) does _not_ depend on kryo, i.e. it's a completely different artifact than kryo (apart from sharing the group id). I.e. even if the kryo5 artifact is produced by depending on kryo, this is not reflected in the created pom. To see this, run `mvn package` in the PR #652 and inspect `target/kryo5-5.0.0-RC3-SNAPSHOT.pom`.

Does it make sense?

**Comment by johnynek** [comment]
Yes. I understand that. What I was discussing was the proposal to have both the versioned and unversioned available as discussed here: https://github.com/EsotericSoftware/kryo/issues/650#issuecomment-466756160

If that is done and an upstream chooses not to use it, which I assume will be common, downstream is stuck. Isn’t this right?

If on the other hand, we only do the versioned artifact then you can have both v5 and earlier on your class path.

**Comment by magro** [comment]
> Yes. I understand that. What I was discussing was the proposal to have both the versioned and unversioned available as discussed here: https://github.com/EsotericSoftware/kryo/issues/650#issuecomment-466756160

I was also discussing the proposal to have both the versioned and unversioned artifact, and more concretely my PR #652.

> If that is done and an upstream chooses not to use it, which I assume will be common, downstream is stuck. Isn’t this right?

I agree to the consequence, but not to the assumption that it will be common - at least not without further restrictions on the context. I'd say it will be common to choose the unversioned artifact for application development, while for libraries the unversioned artifact should be used.

That's the guideline I'd propose for this:
* library authors must/should use the versioned artifact (i.e. as a dependency for a (public) libary)
* application developers can use the unversioned artifact (i.e. as a dependency for an application)

**Comment by masc3d** [comment]
I don't quite understand @magro 

how would a class from versioned artifact be compatible with it's unversioned counterpart as it's in different package?

**Comment by masc3d** [comment]
and why would any legacy consumer actually prefer the unversioned artifact / package?

**Comment by magro** [comment]
> how would a class from versioned artifact be compatible with it's unversioned counterpart as it's in different package?

Why should this be needed?

E.g. spark or chill or whatever library could use the versioned artifact. The application itself could still use the unversioned artifact for other, application level serialization stuff.

> and why would any legacy consumer actually prefer the unversioned artifact / package?

Because it's easier to just update a version than changing (more) code. Maybe it's not _much_ effort to change imports, but it's definitely more. Not sure what you're meaning with "legacy consumer" btw, I interpreted this as "application developer".

**Comment by magro** [comment]
Just to mention it: currently the [Kryo versioning policy](https://github.com/EsotericSoftware/kryo#kryo-versioning-and-upgrading) is to let the major version reflect serialization compatibility and the minor version binary compatibility.
I.e. we'd either have to change the versioning policy or use both major + minor version for the artifact id / package name (e.g. "kryo50").

**Comment by masc3d** [comment]
> Why should this be needed?

depends on the library. as soon as a kryo lib exposes any kryo reference it won't work and has to follow suit by also maintaining / releasing dual artifact / package.

> Maybe it's not _much_ effort to change imports, but it's definitely more.

saving some app devs minimum effort of trivial package name change won't presumably rectify dual artifact / package policy, as it's less clear for any consumer and may also adversely affect lib developers.

**Comment by masc3d** [comment]
> with "legacy consumer"

I was referring to any consumer <= 4 intending to move to v5.

**Comment by magro** [comment]
I agree, in terms of simplicity just providing the versioned artifact is the better way. I'm also fine with that way.

**Comment by masc3d** [comment]
> I.e. we'd either have to change the versioning policy or use both major + minor version for the artifact id

while breaking binary compat is not nice, it's also not a blocker and consumers can deal with it if necessary. 

so if you'd ask me, Id vote for package by major version eg. `kryo5` or `kryo.v5` being sufficient.

**Comment by magro** [comment]
@masc3d As I understood @johnynek, in the context of spark, storm etc broken binary compatibility is an issue (see also [this comment](https://github.com/EsotericSoftware/kryo/issues/650#issuecomment-466815679))...

**Comment by masc3d** [comment]
> because serialization compatibility is not solved that way

not solved automagically, but providing a way for consumers in distributed systems to smoothly transition to v5 instead of being blocked / stuck with v4.

since not every binary compat breakage might be a severe issue for consumers, maybe a good compromise could be to differentiate between minor and major breaking api changes and consolidate the latter with major kryo version / package.

anyhow I would be fine with any approach that doesn't block a way forward.

**Comment by NathanSweet** [comment]
> Because kryo is not recommended for persisting serialized data,

This is definitely not true. Kryo is amazing for persistent data. Migrating to newer Kryo versions can be difficult, but this is not a reason to avoid Kryo for persisting data.

> Maybe we could/should provide an _additional_ artifact "kryo5" which comes with renamed package name.

If we do #652 (unversioned packages + shaded) or we only have versioned packages then the package name should be `kryo.v5` so that everything is grouped under `kryo`, rather than dumping it all under `com.esotericsoftware`.

I'm OK with versioned packages at the source code level. We'll give it a little while if anyone wants to discuss this, otherwise I'll make the change in a a few days.

**Comment by magro** [comment]
> the package name should be `kryo.v5` so that everything is grouped under `kryo`, rather than dumping it all under `com.esotericsoftware`.

In #652 I used `com.esotericsoftware.kryo5` (instead of `...kryo.v5`) for the additional versioned artifact id and JPMS module name because AFAIK JPMS requires module names not to be overlapping (which would have been the case for `kryo` and `kryo.v5`).

> I'm OK with versioned packages at the source code level. We'll give it a little while if anyone wants to discuss this, otherwise I'll make the change in a a few days.

Versioned packages at the source code level is not sufficient IMHO, at least not if we want to address the diamond dependency problem (which is an issue for e.g. spark etc).
To address this, we also need a versioned artifact id (and JPMS module name), and all dependencies must be shaded under a versioned base package.

**Comment by NathanSweet** [comment]
> JPMS requires module names not to be overlapping

Welp, it's not acceptable to use `kryo5`, `kryo6`, etc -- that would make a mess of the `com.esotericsoftware` namespace. Can we just not have a `kryo` module for JPMS? We can have `kryo.v5` and `kryo.v4` (and older, if necessary). If you want to use JPMS you choose one of these, otherwise existing `kryo` artifacts can be used but we don't publish any more.

You're right that versioned packages solve dependency conflicts for Kryo, but not Kryo's dependencies. This can be important when moving to a new Kryo version, eg if the newer Kryo uses a newer ASM.

Shading Kryo and all its dependencies is a bit nasty because it can result in the same classes being loaded twice, eg if other code also uses ASM. It may not be a big deal in most cases, but I don't like forcing it on all users. In the past I believe we provide Kryo as both unshaded and shaded so users can decide.

I change my proposal to leave the source as is and continue to provide Kryo as both unshaded and shaded. If loading multiple Kryo versions without juggling classloaders is important to you or you want to avoid conflicts with Kryo's dependencies (libraries that expose Kryo likely fall under one or both of these) then you would use the shaded version. If you don't care about that, then you can choose to use the unshaded version.

> And a PR world allow us to discuss the single changes...

We should all try to agree on how we want to improve things before considering the exact changes. If we agree with the paragraph above, then it looks like #652 is good except for the artifact ID issue.

**Comment by magro** [comment]
I'm fine with unversioned sources and an additional versioned / shaded artifact.
One thing I just realized is that the versioned shaded jar will come without a suitable sources jar, i.e. the developer experience will be worse in that regard. 

Re naming: Because the artifact id determines the jar name I'd propose s.th. like `kryo5` as the artifact id (under the group id `com.esotericsoftware.kryo`).
Normally I'd directly derive the JPMS module name from the group + artifact id, but I'm not sure what happens if an application uses the unversioned kryo and has a transitive dependency on a versioned kryo. Maybe we choose one way now and invest more time into consideration as soon as somebody submits an issue.

**Comment by magro** [comment]
The versioned, self-contained artifact is now available as `com.esotericsoftware.kryo:kryo5:5.0.0-RC5`.
